Create closure to do sorting.

aldhsu opened the pull request:
Was hacking on #277 and realised that splitting the directories out wasn't necessary so jumped to the sorting with a created closure.

## Future architecture for sorting:
1. Take flags and "compose"(not sure how this works in Rust, so I just went with a loop), sorting functions together to create the sorting function
1. Cache sorting function to reuse for the recursive calls to `Core#sort`

The test pass but I would like to do some performance testing and complete the second step.

aldhsu commented:
# Performance testing
Using the closure:
```
~/Code/lsd/target/release/lsd -R --group-dirs=last  1.93s user 1.47s system 89% cpu 3.802 total
```

Without the closure:
```
lsd -R --group-dirs=last  1.91s user 1.49s system 89% cpu 3.821 total
```

Doesn't look like much has changed.

aldhsu commented:
@Peltoche this is "ready" now. Let me know what I can do to get it merged.

Peltoche commented:
Hi @aldhsu ,

A big PR with a lot of change have been merge. I think we are ready now to merge your PR if you succeed to rebase it on top of the current master

aldhsu commented:
Branch is up to date now.
In the current state, it should be simpler to add new sort criteria,  add support for specifying sort direction on individual fields and multisorting.

Possibly minor effect on performance, recursive sorting around 100k files:
Before
```
Benchmark #1: lsd --recursive -r --sizesort
  Time (mean ± σ):      2.171 s ±  0.015 s    [User: 1.108 s, System: 1.037 s]
  Range (min … max):    2.149 s …  2.187 s    10 runs
```
After, branch built with --release
```
Benchmark #1: /Users/allen.hsu/Code/lsd/target/release/lsd --recursive -r --sizesort
  Time (mean ± σ):      2.261 s ±  0.032 s    [User: 1.187 s, System: 1.042 s]
  Range (min … max):    2.218 s …  2.328 s    10 runs
```

aldhsu commented on src/sort.rs:
```diff
@@ -2,102 +2,63 @@ use crate::flags::{DirOrderFlag, Flags, SortFlag, SortOrder};
+    let mut sorters: Vec<(bool, Sorter)> = vec![];
+    match flags.directory_order {
+        DirOrderFlag::First => {
+            sorters.push((false, Box::new(with_dirs_first)));
```
Addressed in https://github.com/Peltoche/lsd/pull/281/commits/bab0f0e3effa3f00977cf812d22c9e24b2188b3a.

zwpaper commented:
@aldhsu this change looks great, can u help to fix the conflict and make this PR happen?

aldhsu commented:
@zwpaper I've updated the PR. I have a bit more knowledge about Rust now and I realized the closure is actually not necessary since closures are just implicit structs. I switched it to store a vec of function pointers on the Core struct instead of a closure, I think this is a bit easier to work with.

# Before
```
Benchmark #1: lsd -R --group-dirs=last
  Time (mean ± σ):     885.4 ms ±  26.4 ms    [User: 551.8 ms, System: 318.1 ms]
  Range (min … max):   854.0 ms … 938.9 ms    10 runs
```

# After
```
Benchmark #1: ~/Code/lsd/target/release/lsd -R --group-dirs=last
  Time (mean ± σ):     923.2 ms ±  23.5 ms    [User: 596.3 ms, System: 318.9 ms]
  Range (min … max):   909.3 ms … 988.2 ms    10 runs
```
This still definitely makes sorting a bit slower.

zwpaper commented:
@meain @Peltoche could u help to review this PR, this could help developing sorting functionality a lot

meain commented:
I will try to take a look at it some time this week.

meain commented:
Hey, thanks a lot. LGTM.
Sorry for the delay.
